aten/src/ATen/function_wrapper.py

```diff
@@ -637,7 +732,10 @@ def device_guard(option, dispatch_options, dispatch_tensor):
     # For factory methods the `DeviceGuard` is already in the template.
     if option.get('device_guard', True):
         if dispatch_options:
-            return 'const DeviceGuard device_guard({}.device());'.format(dispatch_options['name'])
+            if any(arg['type'] == 'c10::optional<ScalarType>' for arg in option['arguments']):
+                return 'auto dev = device.has_value() ? device.value() : Device(kCPU);\nconst DeviceGuard device_guard(dev);'
```

Comment:
Maybe you should just use `OptionalDeviceGuard` here?

Reply:
I think it would be safe to just unconditionally use OptionalDeviceGuard, it will handle both `optional<Device>` and `Device device` correctly.

Reply:
Well, yes, if you're treating DeviceGuard as a black box, this is something to be concerned about, but I happen to know that `DeviceGuard(kCPU)` is a no-op ;) (Go take a look at the docblock of the class and remind yourself what a DeviceGuard does.)